File: rnamake/unittests/x3dna_unittest.py
import unittest
import os
import shutil
import rnamake.x3dna

class X3dnaUnittest(unittest.TestCase):

    def test_generate_ref_frame(self):
        path = rnamake.settings.UNITTEST_PATH + "resources/motifs/p4p6/p4p6.pdb"
        x3dna = rnamake.x3dna.X3dna()
        x3dna.generate_ref_frame(path)
        if not os.path.isfile("ref_frames.dat"):
            self.fail("ref_frames.dat file should of been generated")
        if os.path.isfile("bestpairs.pdb"):
            self.fail("extra files should of been deleted")

        path = rnamake.settings.UNITTEST_PATH + "resources/motifs/p4p6_2"
        try:
            x3dna.generate_ref_frame(path)
            self.fail()
        except IOError:
            pass
        except:
            self.fail("got an unexpected error")

        # clean up
        if not os.path.isfile("ref_frames.dat"):
            self.fail("did not produce ref_frames.dat")

        os.remove("ref_frames.dat")

    def test_generate_dssr_file(self):
        path = rnamake.settings.UNITTEST_PATH + "resources/motifs/p4p6/p4p6.pdb"
        x3dna = rnamake.x3dna.X3dna()
        x3dna.generate_dssr_file(path)
        if not os.path.isfile("p4p6_dssr.out"):
            self.fail("did not create _dssr.out file")

        # clean up
        if not os.path.isfile("p4p6_dssr.out"):
            self.fail("did not produce p4p6_dssr.out")

        os.remove("p4p6_dssr.out")

    def test_get_basepairs(self):
        path = rnamake.settings.UNITTEST_PATH + "resources/motifs/p4p6/p4p6.pdb"
        x3dna = rnamake.x3dna.X3dna()
        basepairs = x3dna.get_basepairs(path)
        if len(basepairs) != 79:
            self.fail("did not produced the expected number of basepairs")


        #os.remove("ref_frames.dat")
        #os.remove("p4p6_dssr.out")

    def test_get_basepairs_compare(self):
        try:
            import redesign.motif
        except:
            self.skipTest("cannot load old REDESIGN package")

        path = rnamake.settings.UNITTEST_PATH + "resources/motifs/p4p6/p4p6.pdb"
        x3dna = rnamake.x3dna.X3dna()
        basepairs = x3dna.get_basepairs(path)

        try:
            path = "/Users/josephyesselman/projects/REDESIGN/redesign/tests/p4p6"
            m = redesign.motif.Motif(path)
        except:
            self.skipTest("something is wrong cannot load p4p6 with REDESIGN")

        for bp in m.base_pairs:
            found = 0
            for xbp in basepairs:
                if bp.res1.num == xbp.res1.num and bp.res2.num == xbp.res2.num:
                    found = 1
                    break
                if bp.res1.num == xbp.res2.num and bp.res2.num == xbp.res1.num:
                    found = 1
                    break

            if not found:
                print bp.res1,bp.res2
                self.fail()

    def test_x3dna_residues(self):
        res1 = rnamake.x3dna.Residue(103,"A")
        res2 = rnamake.x3dna.Residue(103,"A")
        if res1 != res2:
            self.fail("residue equality is not working")

        res3 = rnamake.x3dna.Residue(102,"A")
        if res1 == res3:
            self.fail("residues should not be equal")

    def test_get_motifs(self):
        path = rnamake.settings.UNITTEST_PATH + "resources/motifs/p4p6/p4p6.pdb"
        x3dna = rnamake.x3dna.X3dna()
        motifs = x3dna.get_motifs(path)

    def test_get_basepairs_local(self):
        path1 = rnamake.settings.UNITTEST_PATH + "resources/motifs/p4p6/p4p6.pdb"
        path = rnamake.settings.UNITTEST_PATH + "p4p6.pdb"
        shutil.copy(path1, path)
        x3dna = rnamake.x3dna.X3dna()
        basepairs = x3dna.get_basepairs(path)
        os.remove("ref_frames.dat")
        os.remove('p4p6_dssr.out')
        os.remove("p4p6.pdb")


def main():
    unittest.main()

if __name__ == '__main__':
    main()

File: rnamake/lib/RNAMake/unittests/util_unittests/x3dna_unittest.cc
//
//  x3dna_unittest.cc
//  RNAMake
//
//  Created by Joseph Yesselman on 5/11/15.
//  Copyright (c) 2015 Joseph Yesselman. All rights reserved.
//

#include "x3dna_unittest.h"
#include "util/file_io.h"


int
X3dnaUnittest::test_generate_ref_frame() {
    String m_path = base_dir() + "/rnamake/unittests/resources/motifs/p4p6_2/p4p6_2.pdb";
    x_.generate_ref_frame(m_path);
    
    if(!(file_exists("ref_frames.dat"))) {
        throw "did not produce ref_frames.dat\n";
    }
       
    if(file_exists("basepairs.pdb")) {
        throw "produced basepairs.pdb which is not needed\n";
    }
    
    //cleanup
    std::remove("ref_frames.dat");
    
    return 1;
}

int
X3dnaUnittest::test_generate_dssr_file() {
    String m_path = base_dir() + "/rnamake/unittests/resources/motifs/p4p6_2/p4p6_2.pdb";
    x_.generate_dssr_file(m_path);
    
    if(!(file_exists("p4p6_2_dssr.out"))) {
        throw "did not produce p4p6_2_dssr.out\n";
    }
    
    std::remove("p4p6_2_dssr.out");
    
    return 1;
}

int
X3dnaUnittest::test_res_compare() {
    X3Residue res1(1, "A", "");
    X3Residue res2(1, "A", "");
    X3Residue res3(2, "A", "");
    X3Residue res4(1, "B", "");

    if(!(res1 == res2)) { return 0; }
    if(res1 == res3)    { return 0; }
    if(res1 == res4)    { return 0; }

    return 1;
}

int
X3dnaUnittest::test_get_basepairs() {
    String m_path = base_dir() + "/rnamake/unittests/resources/motifs/p4p6/p4p6.pdb";
    auto basepairs = x_.get_basepairs(m_path);
    
    if(file_exists("p4p6_dssr.out") || file_exists("ref_frames.dat")) {
        throw UnittestException("produced x3dna files when they already exists");
    }
    
    m_path = base_dir() + "/rnamake/unittests/resources/motifs/p4p6_2/p4p6_2.pdb";
    basepairs = x_.get_basepairs(m_path);

    if(! file_exists("p4p6_2_dssr.out") || ! file_exists("ref_frames.dat")) {
        throw UnittestException("did not produce x3dna files when required");
    }
    
    //cleanup
    std::remove("ref_frames.dat");
    std::remove("p4p6_2_dssr.out");

    return 1;
}

int
X3dnaUnittest::test_get_motifs() {
    String m_path = base_dir() + "/rnamake/unittests/resources/motifs/p4p6/p4p6.pdb";
    auto motifs = x_.get_motifs(m_path);
    
    
    return 1;
}



int
X3dnaUnittest::run() {
    //if (test_generate_ref_frame() == 0)    {  std::cout << "test_generate_ref_frame failed" << std::endl; }
    //if (test_generate_dssr_file() == 0)    {  std::cout << "test_generate_ref_frame failed" << std::endl; }
    //if (test_res_compare() == 0)           {  std::cout << "test_res_compare failed" << std::endl;}
    //if (test_get_basepairs() == 0)         {  std::cout << "test_get_basepairs failed" << std::endl;}
    if (test_get_motifs() == 0)            {  std::cout << "test_get_motifs failed" << std::endl;}

    return 1;
    
}


int
X3dnaUnittest::run_all() {
    String name = "X3dnaUnittest";
    typedef int (X3dnaUnittest::*fptr)();
    std::map<String, fptr> func_map;
    func_map["test_generate_ref_frame" ] = &X3dnaUnittest::test_generate_ref_frame;
    func_map["test_generate_dssr_file" ] = &X3dnaUnittest::test_generate_dssr_file;
    func_map["test_res_compare"        ] = &X3dnaUnittest::test_res_compare;
    func_map["test_get_basepairs"      ] = &X3dnaUnittest::test_get_basepairs;
    
    for(auto const & kv : func_map) {
        try {
            int result = (this->*kv.second)();
            if(result == 0) {
                std::cout << name << "::" << kv.first << " FAILED!" << std::endl;
            }
        }
        catch(...) {
            std::cout << name << "::" << kv.first << " returned ERROR!" << std::endl;
        }
        
    }
    
    return 0;
}





